### PR TITLE
bpo-31993: Do not create frames for large bytes and str objects

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
when serialize into memory buffer with C pickle implementations.

This optimization already is performed when serialize into memory with Python pickle implementations or into a file with both implementations.

https://bugs.python.org/issue31993

### CODE REVIEW

File: Modules/_pickle.c [hunk 1]
```diff
@@ -2276,7 +2291,10 @@ save_bytes(PicklerObject *self, PyObject *obj)
         else {
             /* Bypass the in-memory buffer to directly stream large data
                into the underlying file object. */
-            if (_Pickler_write_large_bytes(self, header, len, obj) < 0) {
+            if (_Pickler_write_large_bytes(self, header, len,
```
Comment (pitrou) [inline Modules/_pickle.c]: What does this change? If `self->write == NULL`, It calls `_Pickler_Write(self, header, len) < 0)` followed by `_Pickler_Write(self, PyBytes_AS_STRING(obj), size)` exactly as above.

Comment (serhiy-storchaka) [reply Modules/_pickle.c]: But it first commits the current frame and temporary disables framing.

File: Modules/_pickle.c [hunk 2]
```diff
@@ -2142,47 +2142,70 @@ save_float(PicklerObject *self, PyObject *obj)
     return 0;
 }
 
-/* No-copy code-path to write large contiguous data directly into the
-   underlying file object, bypassing the output_buffer of the Pickler. */
+/* Write large contiguous data directly into the underlying file object,
```
Comment (pitrou) [inline Modules/_pickle.c]: Can you rephrase the comments now that this is called also when there is no underlying file object?
See comments in `pickle.py` for inspiration.

**Comment by serhiy-storchaka** [comment]
Using `PyMemoryView_FromMemory()` isn't safe since it refers to the temporary data that will be deallocate after destroying a string.

File: Modules/_pickle.c [hunk 3]
```diff
@@ -2142,47 +2142,73 @@ save_float(PicklerObject *self, PyObject *obj)
...
+    int bypass_buffer = (data_size >= FRAME_SIZE_TARGET);
+    int framing = self->framing;
+
+    if (bypass_buffer) {
+        assert(self->output_buffer != NULL);
+        /* Commit the previous frame. */
+        if (_Pickler_CommitFrame(self)) {
+            return -1;
+        }
+        /* Disable frameing temporarily */
```
Comment (ogrisel) [inline Modules/_pickle.c]: typo: framing

File: Modules/_pickle.c [hunk 4]
```diff
@@ -2142,47 +2142,73 @@ save_float(PicklerObject *self, PyObject *obj)
...
+            return -1;
+        }
 
-    /* Reinitialize the buffer for subsequent calls to _Pickler_Write. */
-    if (_Pickler_ClearBuffer(self) < 0) {
-        return -1;
+        /* Stream write the payload into the file without going through the
+           output buffer. */
+        if (payload == NULL) {
+            payload = mem = PyBytes_FromStringAndSize(data, data_size);
```
Comment (ogrisel) [inline Modules/_pickle.c]: This causes an extra memory copy for large `str` (unicode) objects (compared to the unsafe previous code that relied on `PyMemoryView_FromMemory`).

This can be measured with:
https://gist.github.com/ogrisel/0e7b3282c84ae4a581f3b9ec1d84b45a

But I am not sure how to write a safe version of the pickling code that would not cause this extra copy. If it's not possible, the entry for bpo-31993 in the changelog should probably be updated.

**Review by ogrisel — COMMENTED** [review]
The refactoring looks good to me. Maybe we should add a test to check that the number of frames is not too large in a pickled list of large bytes instances. For instance using `count_opcode(pickle.FRAME, pickled)` in `pickletester.py`.

Please also find a comment below in the memory copy induced by `PyBytes_FromStringAndSize` when dumping large unicode objects.

File: Modules/_pickle.c (hunk 4 above)
Comment (pitrou) [reply Modules/_pickle.c]: At least I would like this to go to a separate PR. Please let's not mix concerns.

**Comment by serhiy-storchaka** [comment]
Thank you for your review. `_pickle.dumps()` was not called by corresponding tests. Now test_framing_large_objects covers it.

**Review by ogrisel — APPROVED** [review]
LGTM.

File: Lib/test/pickletester.py [hunk 5]
```diff
@@ -2097,20 +2097,18 @@ def test_framing_large_objects(self):
         N = 1024 * 1024
         obj = [b'x' * N, b'y' * N, 'z' * N]
         for proto in range(4, pickle.HIGHEST_PROTOCOL + 1):
-            for fast in [True, False]:
+            for fast in [True, False, None]:
                 with self.subTest(proto=proto, fast=fast):
-                    if hasattr(self, 'pickler'):
+                    if fast is None:
+                        pickled = self.dumps(obj, proto)
```
Comment (pitrou) [inline Lib/test/pickletester.py]: What is the point of the `None` test? Can you please add a comment?
